src/app/server-cluster/ServerClusterInterface.h

```diff
@@ -0,0 +1,223 @@
+    /// Specifically this WILL NOT be called (and does not need to implement handling for) the
+    /// following attribute IDs:
+    ///     - AcceptedCommandList::Id
+    ///     - AttributeList::Id
+    ///     - GeneratedCommandList::Id
+    virtual DataModel::ActionReturnStatus ReadAttribute(const DataModel::ReadAttributeRequest & request,
+                                                        AttributeValueEncoder & encoder) = 0;
+
+    /// WriteAttribute MUST be done on a valid attribute path. `request.path` is expected to have `GetClusterId` as the cluster
+    /// id as well as an attribute that is included in a `Attributes` call.
+    ///
+    /// Must only be implemented if writable attributes are defined.
+    ///
+    /// Default implementation errors out with an unsupported write on every attribute.
+    virtual DataModel::ActionReturnStatus WriteAttribute(const DataModel::WriteAttributeRequest & request,
+                                                         AttributeValueDecoder & decoder);
+
+    /// Attribute list MUST contain global attributes.
+    ///
+    /// Specifically these attributes MUST always exist in the list for all clusters:
+    ///     - ClusterRevision::Id
+    ///     - FeatureMap::Id
+    ///     - AcceptedCommandList::Id
+    ///     - AttributeList::Id
+    ///     - GeneratedCommandList::Id
+    /// See SPEC 7.13 Global Elements: `Global Attributes` table
+    ///
+    /// Must only be implemented if support for any non-global attributes
+    /// is required.
+    ///
+    /// Default implementation just returns the above global attributes.
+    virtual CHIP_ERROR Attributes(const ConcreteClusterPath & path, DataModel::ListBuilder<DataModel::AttributeEntry> & builder);
+
+    ///////////////////////////////////// Command Support /////////////////////////////////////////////////////////
+
+    /// Must only be implemented if commands are supported by the cluster
+    ///
+    /// Default implementation errors out with an UnspportedCommand error.
+    virtual std::optional<DataModel::ActionReturnStatus>
+    InvokeCommand(const DataModel::InvokeRequest & request, chip::TLV::TLVReader & input_arguments, CommandHandler * handler);
+
+    /// Must only be implemented if commands are supported by the cluster
+    ///
+    /// Default implementation is a NOOP (no list items generated)
+    virtual CHIP_ERROR AcceptedCommands(const ConcreteClusterPath & path,
+                                        DataModel::ListBuilder<DataModel::AcceptedCommandEntry> & builder);
+
+    /// Must only be implemented if commands are supported by the cluster that return values
+    ///
+    /// Default implementation is a NOOP (no list items generated)
+    virtual CHIP_ERROR GeneratedCommands(const ConcreteClusterPath & path, DataModel::ListBuilder<CommandId> & builder);
+
+private:
+    DataVersion mDataVersion; // will be random-initialized as per spec
```

Comment:
I think that these interfaces should be free from functionality and data members.

Requiring a single implementation of data version makes this interface less useful.

- We can't precisely reconstruct specific cluster state for test or other purposes
- It adds unnecessary coupling to the crypto library & we can't inject the entropy source
- We can't inject a "fuzzer" server cluster to try to expose bugs with data version handling.

Keeping the interface pure virtual is the best practice.

Reply:
How about the `IntrusiveList` bit? I would generally want classes from clusters to be directly usable within registries and mocks will be hard to use if they don't have that intrusiveness.

Reply:
It seems like any ember compatibility code should be able to be configured out of the build entirely, since we want to make it possible to have a legacy-free build and soon.

Reply:
I made the base class pure virtual **except** the intrusive list bit since I could not find a good way to minimize RAM usage and keep pure virtual without intrusiveness. Created a "Standard" class that implements the interface with the implementations it had before. GetDataVersion is now pure virtual.

I think API-compatibility can be maintained if we figure out a better way to decouple the registry from interface, for now however I would rather leave that for future work as it seems splittable.